### PR TITLE
refactor: refactor util function for the validator api

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Make the code testable. 

**Description**

- Refactor the util functions to convert to pure functions. 
- Make functions more testable 
- Add more unit tests 
- Refactor the `clock` implementation to make it more generic and testable. 

Related #6159

**Steps to test or reproduce**

- Run all tests

### CODE REVIEW

**Review by g11tech — CHANGES_REQUESTED** [review]
wanna block it to give a deep review shortly

**Comment by nazarhussain** [comment]
> wanna block it to give a deep review shortly

Will merge after your review/approval.

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (twoeths) [inline packages/beacon-node/src/api/impl/validator/index.ts:180]: what's the implication of this change?

File: packages/beacon-node/src/util/clock.ts
Comment (twoeths) [inline packages/beacon-node/src/util/clock.ts:133]: I don't see this function equivalent to old version. Old version returns current slot, or `current slot + 1`. This function return current slot, or `current slot - 1`

File: packages/beacon-node/src/util/clock.ts
Comment (twoeths) [inline packages/beacon-node/src/util/clock.ts:138]: same to the above comment, this function and the old version is not the same

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (nazarhussain) [reply packages/beacon-node/src/api/impl/validator/index.ts:180]: There is no implication apart from making it more readable and testable.

File: packages/beacon-node/src/util/clock.ts
Comment (nazarhussain) [reply packages/beacon-node/src/util/clock.ts:133]: I added detail doc comment explaining the behavior of the function and the all the usage is also aligned with it.

File: packages/beacon-node/src/util/clock.ts
Comment (nazarhussain) [reply packages/beacon-node/src/util/clock.ts:138]: I added detail doc comment explaining the behavior of the function and the all the usage is also aligned with it.

File: packages/beacon-node/src/util/clock.ts
Comment (twoeths) [reply packages/beacon-node/src/util/clock.ts:133]: now the doc says `Return the current slot with future tolerance e.g. currentSlot or currentSlot + 1`

but implementation returns `currentSlot or currentSlot - 1`

```typescript
return toleranceMs > 0 && now > currentSlotStartTime && now - currentSlotStartTime <= toleranceMs
      ? this.currentSlot - 1
      : this.currentSlot;
```

Comment (nazarhussain) [reply packages/beacon-node/src/util/clock.ts:133]: Please suggest how you think we should describe it well; 

```
* -----------------------------------------------------------------------------------
* |      6          |    7                       |           8                 |
* |                 |<-futureToleranceMs->       |
```

The slot time within the range of `<futureToleranceMs>` will be considered current. If the clock passed `6` and now in the `7` slot but within the range of the `futureToleranceMs` then it will return `6`, not the 7. 

e.g. If slot time is 6 seconds and we set future tolerance to the 2 sec, then first 2 second of each slot will still be considered to be previous slot.

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (wemeetagain) [inline packages/beacon-node/src/api/impl/validator/index.ts:128]: are these comments no longer valid? please keep

File: packages/beacon-node/src/util/clock.ts
Comment (wemeetagain) [inline packages/beacon-node/src/util/clock.ts]: this line isn't needed

File: packages/beacon-node/src/util/clock.ts
Comment (wemeetagain) [inline packages/beacon-node/src/util/clock.ts]: the ?? 0 isn't needed

Can also use [`??=`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment) like:

```suggestion
    futureToleranceMs ??= pastToleranceMs;
```

File: packages/beacon-node/src/util/clock.ts
Comment (wemeetagain) [reply packages/beacon-node/src/util/clock.ts:138]: Why is the previous implementation not sufficient?

File: packages/beacon-node/test/unit/api/impl/validator/produceBlockV2.test.ts
Comment (wemeetagain) [inline packages/beacon-node/test/unit/api/impl/validator/produceBlockV2.test.ts]: :eyes:

File: packages/beacon-node/src/chain/validation/attestation.ts
Comment (wemeetagain) [inline packages/beacon-node/src/chain/validation/attestation.ts:491]: Why did you rename / switch these methods?

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (wemeetagain) [inline packages/beacon-node/src/api/impl/validator/index.ts:113]: This seems to throw even if a past slot is passed as an argument. This is a change from the previous behavior.

Comment (wemeetagain) [reply packages/beacon-node/src/api/impl/validator/index.ts:113]: The naming "disparity" is confusing since its being used in subtlely different ways here vs in `currentEpochWithDisparity`.
Here its using `clock.isCurrentSlotGivenTolerance(.., 0, DISPARITY_MS)`
There its using `clock.isCurrentSlotGivenTolerance(.., DISPARITY_MS, 0)`

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: packages/beacon-node/src/util/clock.ts
Comment (nazarhussain) [reply packages/beacon-node/src/util/clock.ts:138]: Previous implementation was working with round `seconds` and truncating the any `ms`. I wanted to make the `clock` implementation consistently work with `ms` everywhere.

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (nazarhussain) [reply packages/beacon-node/src/api/impl/validator/index.ts:113]: Here we are sliding the `clock slot` to the future, in the epoch we are sliding it to the past. I think that was the original behavior. 

If we want to make it consistent please suggest.

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (nazarhussain) [reply packages/beacon-node/src/api/impl/validator/index.ts:128]: No comments are completely removed, either updated or moved to specific utility functions.

File: packages/beacon-node/src/chain/validation/attestation.ts
Comment (nazarhussain) [reply packages/beacon-node/src/chain/validation/attestation.ts:491]: As we really want to have future tolerance, as the peer clock may be behind.

Considering that no tests or test fixture changed, means its actually working right.

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (g11tech) [inline packages/beacon-node/src/api/impl/validator/index.ts:374]: no please don't rename the functions as they have sematics attached to them

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (g11tech) [inline packages/beacon-node/src/api/impl/validator/index.ts:158]: pls dont  mix this change with the other clock changes here

File: packages/beacon-node/src/api/impl/validator/index.ts
Comment (g11tech) [inline packages/beacon-node/src/api/impl/validator/index.ts:542]: hold off on any produceBlock changes except the waitForSlot changes that you are going for here, there are changes stacked on followup PRs which will get conflicted with this anyway

**Comment by nazarhussain** [comment]
Closing as this refactor got outdated.

**Pull request closed without merging**
